agent/http.go

```diff
@@ -161,28 +160,55 @@ func (fs *redirectFS) Open(name string) (http.File, error) {
 	return file, err
 }
 
-type templatedIndexFS struct {
-	fs           http.FileSystem
-	templateVars func() map[string]interface{}
+type settingsInjectedIndexFS struct {
+	fs         http.FileSystem
+	UISettings map[string]interface{}
 }
 
-func (fs *templatedIndexFS) Open(name string) (http.File, error) {
+func (fs *settingsInjectedIndexFS) Open(name string) (http.File, error) {
 	file, err := fs.fs.Open(name)
 	if err != nil || name != "/index.html" {
 		return file, err
 	}
 
-	content, _ := ioutil.ReadAll(file)
+	content, err := ioutil.ReadAll(file)
+	if err != nil {
+		return nil, fmt.Errorf("failed reading index.html: %s", err)
+	}
 	file.Seek(0, 0)
-	t, err := template.New("fmtedindex").Parse(string(content))
+
+	// Replace the placeholder in the meta ENV with the actual UI config settings.
+	// Ember passes the ENV with URL encoded JSON in a meta tag. We are replacing
+	// a key and value that is the encoded version of `"CONSUL_UI_SETTINGS":"__CONSUL_UI_SETTINGS_GO_HERE__"`
+	// with a URL-encoded JSON blob representing the actual config.
+
+	// First built an escaped, JSON blob from the settings passed.
+	bs, err := json.Marshal(fs.UISettings)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed marshalling UI settings JSON: %s", err)
 	}
-	var out bytes.Buffer
-	if err := t.Execute(&out, fs.templateVars()); err != nil {
-		return nil, err
+	// We want to remove the first and last chars which will be the { and } since
+	// we are injecting these variabled into the middle of an existing object.
+	bs = bytes.Trim(bs, "{}")
+
+	// We use PathEscape because we don't want  spaces to be turned into "+" like
+	// QueryEscape does.
+	escaped := url.PathEscape(string(bs))
+
+	content = bytes.Replace(content,
+		[]byte("%22CONSUL_UI_SETTINGS%22%3A%22__CONSUL_UI_SETTINGS_GO_HERE__%22"),
```

Comment:
This doesn't seem to match:

```
CONSUL_UI_SETTINGS_PLACEHOLDER: '__CONSUL_UI_SETTINGS_GO_HERE__'
```

Am I missing something? I guess the quotes could be added by something else, but isn't `PLACEHOLDER` missing?

It might be nice to have a replacement string that is the same literal as the source, but I'm not sure how to do that without it matching twice.

Reply:
If I am not missing something, is this problem being hidden by the fact that we don't validate bindata in CI? I guess this is because previously our workflow was to not update it on every PR, but I'd like to change that so that we can verify that bindata is correct on every PR.
The prior art for this is `check-generated-protobuf` and `check-vendor`

Reply:
Yes good catch, I changed the name when writing up the comments and didn't fix it right through. I'll fix that. This tested correctly because I also didn't rebuild bindata after changing the JS side...

Improvements to cover more of this plumbing in CI is certainly a good goal! I'm not sure it would have caught this without a full end-to-end UI test which is another level of complexity again but one day hopefully will exist!

Reply:
I thought we did have CI to check that the assetfs was rebuilt if there were UI changes actually 🤔 perhaps we talked about that but didn't add it yet. Do you recall @alvin-huang, @johncowen?

Reply:
> I'm not sure it would have caught this without a full end-to-end UI test 

True, It wouldn't have caught the bug, but it would have failed CI because the source file did not match the generated bindata file, which would be a good indicator that something was wrong. Fixing the CI failure would fix the bug (by generating the file again).